core/src/com/unciv/logic/map/TileInfo.kt

```diff
@@ -68,49 +68,47 @@ open class TileInfo {
         return false
     }
 
-    fun containsUnique(unique: String): Boolean {
-        return isNaturalWonder() && getNaturalWonder().uniques.contains(unique)
-    }
+    fun containsUnique(unique: String): Boolean =
+            isNaturalWonder() && getNaturalWonder().uniques.contains(unique)
     //region pure functions
 
     /** Returns military, civilian and air units in tile */
     fun getUnits(): List<MapUnit> {
-        if(militaryUnit==null && civilianUnit==null && airUnits.isEmpty())
+        if (militaryUnit == null && civilianUnit == null && airUnits.isEmpty())
             return emptyList() // for performance reasons - costs much less to initialize than an empty ArrayList or list()
-        val list = ArrayList<MapUnit>(2)
-        if(militaryUnit!=null) list.add(militaryUnit!!)
-        if(civilianUnit!=null) list.add(civilianUnit!!)
+        val list = ArrayList<MapUnit>(2 + airUnits.size)
+        if (militaryUnit != null) list.add(militaryUnit!!)
+        if (civilianUnit != null) list.add(civilianUnit!!)
         list.addAll(airUnits)
         return list
     }
 
     fun getCity(): CityInfo? = owningCity
 
-    fun getLastTerrain(): Terrain = if (terrainFeature != null) getTerrainFeature()!! else if(naturalWonder != null) getNaturalWonder() else getBaseTerrain()
+    fun getLastTerrain(): Terrain = terrainFeature?.let { ruleset.terrains.getValue(it) }
+            ?: naturalWonder?.let {ruleset.terrains.getValue(it)} ?: getBaseTerrain()
 
     fun getTileResource(): TileResource =
             if (resource == null) throw Exception("No resource exists for this tile!")
             else ruleset.tileResources[resource!!]!!
 
+    fun getTileResourceOrNull(): TileResource? = resource?.let { ruleset.tileResources[it] }
```

Comment:
I know your guide says to use less `let` but these seem like good places for this